pallets/referrals/src/lib.rs

```diff
@@ -0,0 +1,554 @@
+								}
+							}
+						}
+					}
+				}
+			});
+
+			Self::deposit_event(Event::Claimed { who, rewards });
+			Ok(())
+		}
+
+		/// Set asset tier reward percentages
+		///
+		/// /// Parameters:
+		/// - `asset_id`: asset id
+		/// - `level`: level
+		/// - `referrer`: referrer percentage that goes to the referrer.
+		/// - `trader`: trader percentage that goes back to the trader.
+		///
+		/// Emits `TierRewardSet` event when successful.
+		#[pallet::call_index(4)]
+		#[pallet::weight(<T as Config>::WeightInfo::set_reward_percentage())]
+		pub fn set_reward_percentage(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			level: Level,
+			referrer: Permill,
+			trader: Permill,
+		) -> DispatchResult {
+			T::AuthorityOrigin::ensure_origin(origin)?;
+
+			//ensure that total percentage does not exceed 100%
+			ensure!(
+				referrer.checked_add(&trader).is_some(),
+				Error::<T>::IncorrectRewardPercentage
+			);
+
+			AssetTier::<T>::mutate(asset_id, level, |v| {
+				*v = Some(Tier { referrer, trader });
+			});
+			Self::deposit_event(Event::TierRewardSet {
+				asset_id,
+				level,
+				referrer,
+				trader,
+			});
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	//TODO: when added to runtime, make sure the account is added to the whitelist of account that cannot be dusted
+	pub fn pot_account_id() -> T::AccountId {
+		T::PalletId::get().into_account_truncating()
+	}
+
+	pub(crate) fn normalize_code(code: ReferralCode<T::CodeLength>) -> ReferralCode<T::CodeLength> {
+		let r = code.into_inner().iter().map(|v| v.to_ascii_uppercase()).collect();
+		ReferralCode::<T::CodeLength>::truncate_from(r)
+	}
+
+	/// Process trader fee
+	/// `source`: account to take the fee from
+	/// `trader`: account that does the trade
+	///
+	/// Returns unused amount on success.
+	#[transactional]
+	pub fn process_trade_fee(
+		source: T::AccountId,
+		trader: T::AccountId,
+		asset_id: T::AssetId,
+		amount: Balance,
+	) -> Result<Balance, DispatchError> {
+		// Does trader have a linked referral account ?
+		let Some(ref_account) = Self::linked_referral_account(&trader) else {
+			return Ok(amount);
+		};
+		// What is the referer level?
+		let Some((level,_)) = Self::referrer_level(&ref_account) else {
+			// Should not really happen, the ref entry should be always there.
+			return Ok(amount);
+		};
+
+		// What is asset fee for this level? if any.
+		let Some(tier) = Self::asset_tier(asset_id, level) else {
+			return Ok(amount);
+		};
+
+		let Some(price) = T::SpotPriceProvider::spot_price(T::RewardAsset::get(), asset_id) else {
+			// no price, no fun.
+			return Ok(amount);
+		};
+
+		let referrer_reward = tier.referrer.mul_floor(amount);
+		let trader_reward = tier.trader.mul_floor(amount);
+		let total_taken = referrer_reward.saturating_add(trader_reward);
+		ensure!(total_taken <= amount, Error::<T>::IncorrectRewardCalculation);
+		T::Currency::transfer(asset_id, &source, &Self::pot_account_id(), total_taken, true)?;
```

Comment:
What happen if `total_taken` is less than ED and `pot` is empty? non-dustable whitelist doesn't work for pallet-balances and HDX.

Reply:
Right... good point. let me think about this one

Reply:
we added a seed amount.